spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnVectorBuilder.java

```diff
@@ -46,8 +46,8 @@ public ColumnVector build(VectorHolder holder, int numRows) {
       } else {
         throw new IllegalStateException("Unknown dummy vector holder: " + holder);
       }
-    } else if (rowIdMapping != null) {
-      return new ColumnVectorWithFilter(holder, rowIdMapping);
+    } else if (withDelete) {
```

Comment:
The new approach is to load all data vectors first, and then apply delete logic to the ColumnarBatch by mutating all the ColumnVectors in place via setRowIdMapping. Initially, all the rowIdMapping are null, and later, the rowIdMapping will be computed and set on all the ColumnVectors. Since the rowIdMapping are set later, at this point, the rowIdMapping is null, we can't use `rowIdMapping != null` to decide to whether to construct a ColumnVectorWithFilter or regular `IcebergArrowColumnVector`. Instead, `withDelete` (which is computed using `deletes != null` in ColumnarBatchReader.readDataToColumnVectors`) is used.